File: fitness/api/routers/journal.py
import asyncio
from datetime import date
from fastapi import APIRouter, Request, Query, HTTPException

from fitness.api.config import (
    JOUR_DIR, SESS_ROOT,
    _uid_from_request, _read_json, logger
)
from db.schemas import JournalResponse
from fitness.firestore.mirror import mirror_journal
from fitness.runtime.journal_store import entries_path, freetext_body, upsert_entry

router = APIRouter()

def _today() -> str:
    return date.today().isoformat()

def _require_uid(request: Request) -> str:
    uid = _uid_from_request(request)
    if not uid or uid == "default":
        raise HTTPException(400, detail="uid Pflicht. Aktive UID fehlt; via ?uid=... oder X-User-UID Header senden.")
    return uid

def _journal_dir(uid: str):
    return SESS_ROOT / uid / "journal"

@router.get("/journal")
def journal_get(request: Request, date_: str = Query(None, alias="date")):
    uid = _uid_from_request(request)
    day = date_ or _today()
    # Pro-uid-Ordner zuerst (mehrere echte Klienten teilen sich sonst denselben
    # globalen JOUR_DIR und überschreiben sich gegenseitig für dasselbe Datum)
    # — JOUR_DIR bleibt als Legacy-Fallback für Einträge von vor diesem Fix.
    for base in (_journal_dir(uid), JOUR_DIR):
        md = base / f"{day}.md"
        jsonl = entries_path(md)
        # JSONL ist die SOT: die editierbare Textarea bekommt nur den
        # Freitext-Eintrag zurück, nicht die gerenderte .md (die enthält
        # zusätzlich die Habit-/Firestore-Blöcke mit <!-- entry:… -->-Markern).
        if jsonl.exists():
            return {"ok": True, "content": freetext_body(jsonl, f"freetext-{day}"),
                    "mtime": jsonl.stat().st_mtime}
        if md.exists():  # noch nicht migrierte .md → roh wie bisher
            return {"ok": True, "content": md.read_text(encoding="utf-8"), "mtime": md.stat().st_mtime}
    raise HTTPException(404)

@router.post("/journal")
async def journal_post(request: Request, date_: str = Query(None, alias="date")):
    uid     = _require_uid(request)
    day     = date_ or _today()
    body    = await request.json()
    content = body.get("content", "")
    dir_ = _journal_dir(uid)
    dir_.mkdir(parents=True, exist_ok=True)
    # Freitext-Tagesblock als EIN Eintrag upserten (stabile id `freetext-<date>`)
    # statt die Datei zu overwriten — ein Daemon-Event (Habit/Firestore-Journal)
    # kann den Freitext so nicht mehr wegräumen und umgekehrt.
    upsert_entry(dir_ / f"{day}.md",
                 {"id": f"freetext-{day}", "source": "freetext", "ts": "", "body": content})
    asyncio.get_event_loop().run_in_executor(None, mirror_journal, day, {"text": content}, uid)
    logger.info(f"journal saved  {uid}/{day}  {len(content)}ch")
    return {"ok": True}

@router.get("/journal/list")
def journal_list(request: Request):
    uid = _uid_from_request(request)
    dirs = [d for d in (_journal_dir(uid), JOUR_DIR) if d.exists()]
    seen: dict[str, float] = {}
    for dir_ in dirs:
        for f in dir_.glob("*.md"):
            mtime = f.stat().st_mtime
            if f.stem not in seen or mtime > seen[f.stem]:
                seen[f.stem] = mtime
    entries = [
        {"date": day, "mtime": mtime}
        for day, mtime in sorted(seen.items(), key=lambda kv: kv[0], reverse=True)[:50]
    ]
    return {"ok": True, "entries": entries}

File: server/routes/journal.mjs
import { z } from "@hono/zod-openapi";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { defineJsonRoute } from "../lib/routes.mjs";

export function registerJournal(app, ctx) {
  const { FITNESS_UID, DATA_DIR, localToday, mirrorJournal, readJournalFull, listJournals, journalEntriesPath, journalFreetextBody, journalUpsertEntry } = ctx;
// ── Journal ───────────────────────────────────────────────────────────────────
app.openapi(defineJsonRoute({
  method: "get",
  path: "/journal",
  tags: ["journal"],
  summary: "Journal-Eintrag lesen",
  query: z.object({ date: z.string().optional() }),
}), async (c) => {
  const { date: dateQ } = c.req.valid("query");
  const uid  = c.req.header("X-User-UID") || FITNESS_UID;
  const date = dateQ || localToday();
  // Firestore-first
  const fsContent = await readJournalFull(uid, date);
  if (fsContent) return c.json({ ok: true, content: fsContent, mtime: date, source: "firestore" });
  // Offline-Fallback: lokale .md Dateien
  const localDirs = [
    { file: path.join(os.homedir(), ".aos", "fitness", "users", uid, "journal", `${date}.md`), label: null },
    { file: path.join(DATA_DIR, "journal", `${date}.md`), label: null },
    { file: path.join(os.homedir(), ".aos", "fuel", "users", uid, "nutrition_journal", `${date}.md`), label: "Fuel" },
  ].filter(({ file }) => fs.existsSync(file));
  if (!localDirs.length) return c.json({ ok: false }, 404);
  const content = localDirs.map(({ file, label }) => {
    // Fitness-Journale: JSONL ist SOT → nur den Freitext-Eintrag in die
    // editierbare Textarea (die gerenderte .md enthält zusätzlich die
    // Habit-/Firestore-Blöcke). Fuel bleibt Rohtext.
    const jsonl = journalEntriesPath(file);
    const text = (label !== "Fuel" && fs.existsSync(jsonl))
      ? journalFreetextBody(file, `freetext-${date}`)
      : fs.readFileSync(file, "utf8");
    return label ? `## ${label} – ${date}\n\n${text}` : text;
  }).join("\n\n---\n\n");
  const mtime = localDirs.map(({ file }) => fs.statSync(file).mtime).reduce((a, b) => a > b ? a : b).toISOString().slice(0, 10);
  return c.json({ ok: true, content, mtime, source: "local" });
});

app.openapi(defineJsonRoute({
  method: "post",
  path: "/journal",
  tags: ["journal"],
  summary: "Journal-Eintrag speichern",
  query: z.object({ date: z.string().optional() }),
  jsonBody: z.object({ content: z.string().optional() }).loose(),
}), async (c) => {
  const { date: dateQ } = c.req.valid("query");
  const { content } = c.req.valid("json");
  const uid           = c.req.header("X-User-UID") || FITNESS_UID;
  const date          = dateQ || localToday();
  // Pro-uid-Ordner (wie GET oben schon macht) statt fix an den beim Server-
  // Start aufgelösten DATA_DIR (= FITNESS_UID) — sonst landet der Eintrag
  // eines anderen Klienten (X-User-UID-Header) immer im falschen Journal.
  const dir           = path.join(os.homedir(), ".aos", "fitness", "users", uid, "journal");
  fs.mkdirSync(dir, { recursive: true });
  const file          = path.join(dir, `${date}.md`);
  // Freitext-Tagesblock als EIN Eintrag (stabile id `freetext-<date>`) in die
  // JSONL-SOT upserten statt die Datei zu overwriten — Daemon-Events und
  // Freitext räumen sich so nicht mehr gegenseitig weg.
  journalUpsertEntry(file, { id: `freetext-${date}`, source: "freetext", ts: "", body: content || "" });
  mirrorJournal(date, { text: content || "" }, uid);
  return c.json({ ok: true });
});

app.openapi(defineJsonRoute({
  method: "get",
  path: "/journal/list",
  tags: ["journal"],
  summary: "Journal-Liste laden",
  query: z.object({ limit: z.coerce.number().int().positive().max(500).optional().default(50) }),
}), async (c) => {
  const { limit: limitCount } = c.req.valid("query");
  const uid = c.req.header("X-User-UID") || FITNESS_UID;
  // Firestore-first
  const fsEntries = await listJournals(uid, limitCount);
  if (fsEntries) return c.json({ ok: true, entries: fsEntries, source: "firestore" });
  // Offline-Fallback
  const dirs = [
    path.join(os.homedir(), ".aos", "fitness", "users", uid, "journal"),
    path.join(DATA_DIR, "journal"),
    path.join(os.homedir(), ".aos", "fuel", "users", uid, "nutrition_journal"),
  ];
  const seen = new Map();
  for (const dir of dirs) {
    if (!fs.existsSync(dir)) continue;
    for (const f of fs.readdirSync(dir).filter(f => f.endsWith(".md"))) {
      const date = f.replace(".md", "");
      const mtime = fs.statSync(path.join(dir, f)).mtime.toISOString();
      if (!seen.has(date) || mtime > seen.get(date)) seen.set(date, mtime);
    }
  }
  const entries = [...seen.entries()]
    .sort((a, b) => b[0].localeCompare(a[0])).slice(0, limitCount)
    .map(([date, mtime]) => ({ date, mtime }));
  return c.json({ ok: true, entries, source: "local" });
});

}
